eth/tracers/live/supply.go

```diff
@@ -0,0 +1,266 @@
+	fmt.Printf("------------------------------\n\n")
+}
+
+func (s *Supply) OnGenesisBlock(b *types.Block, alloc core.GenesisAlloc) {
+	if s.hasGenesisProcessed {
+		return
+	}
+
+	s.resetDelta()
+
+	s.delta.Number = b.NumberU64()
+	s.delta.Hash = b.Hash()
+	s.delta.ParentHash = b.ParentHash()
+
+	delta := big.NewInt(0)
+
+	// Initialize supply with total allocation in genesis block
+	for _, account := range alloc {
+		delta.Add(delta, account.Balance)
+	}
+
+	s.delta.Delta = delta
+
+	s.hasGenesisProcessed = true
+
+	out, _ := json.Marshal(s.delta)
+	s.logger.Println(string(out))
+
+	fmt.Printf("OnGenesisBlock:\n\t --[supply] %s\n\n", out)
+}
+
+func (s *Supply) OnBalanceChange(a common.Address, prevBalance, newBalance *big.Int, reason state.BalanceChangeReason) {
+	diff := new(big.Int).Sub(newBalance, prevBalance)
+
+	switch reason {
+	case state.BalanceIncreaseGenesisBalance:
+		s.delta.Delta.Add(s.delta.Delta, diff)
+	case state.BalanceIncreaseRewardMineUncle:
+	case state.BalanceIncreaseRewardMineBlock:
+		s.delta.Reward.Add(s.delta.Reward, diff)
+	case state.BalanceIncreaseWithdrawal:
+		s.delta.Withdrawals.Add(s.delta.Withdrawals, diff)
+	case state.BalanceDecreaseSelfdestructBurn:
+		s.delta.Burn.Sub(s.delta.Burn, diff)
+	default:
+		// fmt.Printf("~~\tNo need to take action. Change reason: %v\n\n", reason)
+		return
+	}
+
+	s.delta.Delta.Add(s.delta.Delta, diff)
+
+	fmt.Printf("\nOnBalanceChange: a=%v, prev=%v, new=%v, \n--\tdiff=%v, reason=%v\n", a, prevBalance, newBalance, diff, reason)
+}
+
+// CaptureStart implements the EVMLogger interface to initialize the tracing operation.
+func (s *Supply) CaptureStart(from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int) {
+	s.txCallstack = make([]supplyTxCallstack, 1)
+
+	s.txCallstack[0] = supplyTxCallstack{
+		calls: make([]supplyTxCallstack, 0),
+	}
+}
+
+// CaptureEnd is called after the call finishes to finalize the tracing.
+func (s *Supply) CaptureEnd(output []byte, gasUsed uint64, err error, reverted bool) {
+	// No need to handle Burned amount if transaction is reverted
+	if !reverted {
+		s.interalTxsHandler(&s.txCallstack[0])
+	}
+}
+
+// interalTxsHandler handles internal transactions burned amount
+func (s *Supply) interalTxsHandler(call *supplyTxCallstack) {
+	// Handle Burned amount
+	if call.burn != nil {
+		s.delta.Burn.Add(s.delta.Burn, call.burn)
+		s.delta.Delta.Sub(s.delta.Delta, call.burn)
+	}
+
+	if len(call.calls) > 0 {
+		// Recursivelly handle internal calls
+		for _, call := range call.calls {
+			callCopy := call
+			s.interalTxsHandler(&callCopy)
+		}
+	}
+}
+
+// CaptureEnter is called when EVM enters a new scope (via call, create or selfdestruct).
+func (s *Supply) CaptureEnter(typ vm.OpCode, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int) {
+	call := supplyTxCallstack{
+		calls: make([]supplyTxCallstack, 0),
+	}
+
+	// TODO shall we check if London enabled or b.BaseFee() != nil
```

Comment:
@s1na I am thinking that I have to check if BaseFee is set and otherwise don't count the calue as Burned. Is this the case?